codice/ddf
# [2.19.x] DDF-UI-248 Update UI automatically on share/unshare/delete

backport of https://github.com/codice/ddf-ui/pull/295

#### What does this PR do?
Add Server-Sent Event messaging to DDF, and uses it to automatically update search forms, workspaces, and result forms on share, unshare, and delete operations 

#### Who is reviewing it? 


@andrewzimmer @bdeining @hayleynorton @zta6 @abel-connexta 

#### Select relevant component teams: 


@codice/ui 

#### Ask 2 committers to review/merge the PR and tag them here.

@pklinef 
@rymach 

#### How should this be tested?


### Setup
* build 
* before starting ddf go into /etc/ directory, update users.attributes file to include
```
"test" : {
       "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress":"test@localhost.local",
       "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/role": ["manager","viewer","systembundles"]
   }
```
and in users.properties files add
`test=test,group,manager,viewer,systembundles,ssh`

* start ddf, profile install, and start the dev server
* log into the test user on `localhost:8993/search/catalog`
* open another window and go to `localhost:8080/search/catalog`

### Share/Unshare
0. Navigate to the workspaces collection page on `admin` user's instance
1. Create a new workspace on `test` user's instance
2. Share the workspace with the `admin` group (should be there already)
3. Verify on the `admin` instance that `test` user's workspace shows up automatically (takes 3 seconds)
4. Share the other way around (from `admin` to `test@localhost.local`) and verify it shows up on `test`'s workspace collection automatically

Repeat these steps for unshare. The workspace should disappear from the user to whom the workspace was unshared.

### Delete
0-3. Do steps 0-3 from *Share/Unshare*
4. Delete the workspace on `test` user's page
5. Verify on the `admin` instance that `test` user's workspace disappears automatically (takes 3 seconds)
6. Delete the other way around and verify

Repeat these steps for search forms and result forms.

### Workspace Search
0. Open a workspace on `admin`'s instance and Click on the Search DDF button
1. Create a new search form on `test` user's instance
2. Share the search form with the `admin` group (should be there already)
3. Verify on the `admin` instance that `test` user's search form shows up in the three-dot menu drop down when you click "Use Other Search Form" (takes 3 seconds)
4. Create a new result form on the `test` user's instance
5. Share that result form with `admin`
6. Verify in `admin` user's instance that the result form shows up in the Result Form dropdown in the advanced search from.  


#### Any background context you want to provide?

#### What are the relevant tickets?
Fixes: https://github.com/codice/ddf-ui/issues/248

#### Screenshots


#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
